plugins/benchmark_test.go

```diff
@@ -0,0 +1,322 @@
+package plugins
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+)
+
+type BenchmarkScenario struct {
+	Accounts          int
+	Regions           int
+	Tables            int
+	ChildrenPerTable  int
+	Columns           int
+	ColumnResolvers   int // number of columns with custom resolvers
+	ResourcesPerTable int
+	ResourcesPerPage  int
+	ResolverMin       time.Duration
+	ResolverStdDev    time.Duration
+	ResolverMean      time.Duration
+	Concurrency       uint64
+}
+
+func (s *BenchmarkScenario) SetDefaults() {
+	if s.Accounts == 0 {
+		s.Accounts = 1
+	}
+	if s.Regions == 0 {
+		s.Regions = 1
+	}
+	if s.Tables == 0 {
+		s.Tables = 1
+	}
+	if s.Columns == 0 {
+		s.Columns = 10
+	}
+	if s.ResourcesPerTable == 0 {
+		s.ResourcesPerTable = 100
+	}
+	if s.ResourcesPerPage == 0 {
+		s.ResourcesPerPage = 10
+	}
+	if s.ResolverMin == 0 {
+		s.ResolverMin = time.Millisecond
+	}
+	if s.ResolverStdDev == 0 {
+		s.ResolverStdDev = 100 * time.Millisecond
+	}
+	if s.ResolverMean == 0 {
+		s.ResolverMean = 10 * time.Millisecond
+	}
+}
+
+type SourceBenchmark struct {
+	*BenchmarkScenario
+
+	b      *testing.B
+	tables []*schema.Table
+	plugin *SourcePlugin
+
+	apiCalls atomic.Int64
+}
+
+func NewSourceBenchmark(b *testing.B, scenario BenchmarkScenario) *SourceBenchmark {
+	scenario.SetDefaults()
+	sb := &SourceBenchmark{
+		BenchmarkScenario: &scenario,
+		b:                 b,
+		tables:            nil,
+		plugin:            nil,
+	}
+	sb.setup(b)
+	return sb
+}
+
+func (s *SourceBenchmark) setup(b *testing.B) {
+	tableResolver := func(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
+		s.simulateAPICall(s.ResolverMin, s.ResolverStdDev, s.ResolverMean)
+		total := 0
+		for total < s.ResourcesPerTable {
+			num := min(s.ResourcesPerPage, s.ResourcesPerTable-total)
+			resources := make([]struct {
+				Column1 string
+			}, num)
+			for i := 0; i < num; i++ {
+				resources[i] = struct {
+					Column1 string
+				}{
+					Column1: "test-table",
+				}
+			}
+			res <- resources
+			total += num
+		}
+		return nil
+	}
+	preResourceResolver := func(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource) error {
+		s.simulateAPICall(s.ResolverMin, s.ResolverStdDev, s.ResolverMean)
+		resource.Item = struct {
+			Column1 string
+		}{
+			Column1: "test-pre",
+		}
+		return nil
+	}
+	columnResolver := func(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource, c schema.Column) error {
+		s.simulateAPICall(s.ResolverMin, s.ResolverStdDev, s.ResolverMean)
+		err := resource.Set(c.Name, "test")
```

Comment:
why not just `return resource.Set(...`

Reply:
meh, yeah. Was just trying to satisfy the linter, this should never really return an error

Reply:
Updated 👍